File: Data Structures & Algorithms/coin-change-ii/submission-7.py
class Solution:
    def change(self, amount: int, coins: List[int]) -> int: 
        memo = {}
        def dfs(i, cur):  
            if (i, cur) in memo: 
                return memo[(i, cur)] 

            if cur > amount or i == len(coins): 
                return 0
            
            if cur == amount:  
                return 1

            memo[(i, cur)] = dfs(i + 1, cur) + dfs(i, cur + coins[i]) 
            return memo[(i, cur)] 

        
        return dfs(0, 0)

File: Data Structures & Algorithms/min-cost-climbing-stairs/submission-7.js
class Solution {
    /**
     * @param {number[]} cost
     * @return {number}
     */
    minCostClimbingStairs(cost) {   
        return Math.min(this.dfs(0, cost), this.dfs(1, cost))
    }  

    dfs(pos, cost) {   
        if (pos >= cost.length) return 0 
        const cur = cost[pos]
        return cur + Math.min(this.dfs(pos + 1, cost), this.dfs(pos + 2, cost))
    }
}
